File: swift/Tests/JSONStructureTests/RelationsAndUcumUnitValidationTests.swift
import XCTest
@testable import JSONStructure

final class RelationsAndUcumUnitValidationTests: XCTestCase {
    func testValidNumericTypeWithUcumUnit() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))
        let schema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:ucum-number",
            "name": "Length",
            "$uses": ["JSONStructureUnits"],
            "type": "number",
            "ucumUnit": "m"
        ]

        let result = validator.validate(schema)
        XCTAssertTrue(result.isValid, "Expected valid schema, got errors: \(result.errors)")
    }

    func testValidNumericTypeWithUnitAndUcumUnit() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))
        let schema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:ucum-both",
            "name": "Length",
            "$uses": ["JSONStructureUnits"],
            "type": "number",
            "unit": "meter",
            "ucumUnit": "m"
        ]

        let result = validator.validate(schema)
        XCTAssertTrue(result.isValid, "Expected valid schema, got errors: \(result.errors)")
    }

    func testValidExtendedNumericTypesWithUcumUnit() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))

        for type in ["int32", "float", "double", "decimal"] {
            let schema: [String: Any] = [
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "urn:example:ucum-\(type)",
                "name": "\(type)WithUcumUnit",
                "$uses": ["JSONStructureUnits"],
                "type": type,
                "ucumUnit": "m"
            ]

            let result = validator.validate(schema)
            XCTAssertTrue(result.isValid, "Expected valid schema for \(type), got errors: \(result.errors)")
        }
    }

    func testInvalidUcumUnitScenariosArePending() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))

        let invalidTypeSchema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:ucum-invalid-type",
            "name": "BadUcumType",
            "type": "string",
            "ucumUnit": "m"
        ]

        let invalidValueSchema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:ucum-invalid-value",
            "name": "BadUcumValue",
            "$uses": ["JSONStructureUnits"],
            "type": "number",
            "ucumUnit": 5
        ]

        let invalidTypeResult = validator.validate(invalidTypeSchema)
        XCTAssertFalse(invalidTypeResult.isValid)
        XCTAssertTrue(invalidTypeResult.errors.contains { $0.message.contains("JSONStructureUnits extension") })
        XCTAssertTrue(invalidTypeResult.errors.contains { $0.message.contains("can only appear in numeric schemas") })

        let invalidValueResult = validator.validate(invalidValueSchema)
        XCTAssertFalse(invalidValueResult.isValid)
        XCTAssertTrue(invalidValueResult.errors.contains { $0.message.contains("'ucumUnit' must be a string.") })
    }

    func testValidRelationsIdentityArray() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))
        let schema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:relations-identity",
            "name": "OrderIdentity",
            "$uses": ["JSONStructureRelations"],
            "type": "object",
            "properties": [
                "id": ["type": "string"],
                "tenantId": ["type": "string"]
            ],
            "identity": ["id", "tenantId"]
        ]

        let result = validator.validate(schema)
        XCTAssertTrue(result.isValid, "Expected valid schema, got errors: \(result.errors)")
    }

    func testValidRelationsDeclarations() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))
        let schema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:relations-valid",
            "name": "OrderRelations",
            "$uses": ["JSONStructureRelations"],
            "type": "object",
            "properties": [
                "id": ["type": "string"],
                "customerId": ["type": "string"],
                "itemIds": [
                    "type": "array",
                    "items": ["type": "string"]
                ],
                "qualifier": ["type": "string"]
            ],
            "relations": [
                "customer": [
                    "cardinality": "single",
                    "targettype": ["$ref": "#/definitions/Customer"]
                ],
                "items": [
                    "cardinality": "multiple",
                    "targettype": ["$ref": "#/definitions/Item"],
                    "scope": "line-items"
                ],
                "qualifiedCustomer": [
                    "cardinality": "single",
                    "targettype": ["$ref": "#/definitions/Customer"],
                    "qualifiertype": ["$ref": "#/definitions/RelationQualifier"]
                ]
            ],
            "definitions": [
                "Customer": [
                    "name": "Customer",
                    "type": "object",
                    "properties": ["id": ["type": "string"]]
                ],
                "Item": [
                    "name": "Item",
                    "type": "object",
                    "properties": ["id": ["type": "string"]]
                ],
                "RelationQualifier": [
                    "name": "RelationQualifier",
                    "type": "string"
                ]
            ]
        ]

        let result = validator.validate(schema)
        XCTAssertTrue(result.isValid, "Expected valid schema, got errors: \(result.errors)")
    }

    func testInvalidRelationsScenariosArePending() throws {
        let validator = SchemaValidator(options: SchemaValidatorOptions(extended: true))

        let schema: [String: Any] = [
            "$schema": "https://json-structure.org/meta/extended/v0/#",
            "$id": "urn:example:relations-invalid",
            "name": "BadRelations",
            "type": "string",
            "identity": ["id"],
            "relations": [
                "customer": [
                    "cardinality": "many",
                    "targettype": ["type": "object"],
                    "scope": ["tenant", 3],
                    "qualifiertype": ["type": "string"]
                ]
            ]
        ]

        let result = validator.validate(schema)
        XCTAssertFalse(result.isValid)
        XCTAssertTrue(result.errors.contains { $0.message.contains("JSONStructureRelations extension") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'identity' can only appear in object or tuple schemas.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'identity' references property 'id' that is not in 'properties'.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'relations' can only appear in object or tuple schemas.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'targettype' must be an object with '$ref'.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'cardinality' must be 'single' or 'multiple'.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'scope' array items must be strings.") })
        XCTAssertTrue(result.errors.contains { $0.message.contains("'qualifiertype' must be an object with '$ref'.") })
    }
}

File: java/src/test/java/org/json_structure/validation/RelationsAndUcumUnitValidationTests.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package org.json_structure.validation;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;

class RelationsAndUcumUnitValidationTests {

    private SchemaValidator validator;

    @BeforeEach
    void setUp() {
        validator = new SchemaValidator();
    }

    @Test
    @DisplayName("Valid numeric type with ucumUnit")
    void validNumericTypeWithUcumUnit() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/ucum-number",
                "name": "Length",
                "$uses": ["JSONStructureUnits"],
                "type": "number",
                "ucumUnit": "m"
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
        assertThat(result.getErrors()).isEmpty();
    }

    @Test
    @DisplayName("Valid numeric type with unit and ucumUnit")
    void validNumericTypeWithUnitAndUcumUnit() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/ucum-both",
                "name": "Length",
                "$uses": ["JSONStructureUnits"],
                "type": "number",
                "unit": "meter",
                "ucumUnit": "m"
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
        assertThat(result.getErrors()).isEmpty();
    }

    @ParameterizedTest
    @ValueSource(strings = {"int32", "float", "double", "decimal"})
    @DisplayName("Valid extended numeric types with ucumUnit")
    void validExtendedNumericTypesWithUcumUnit(String typeName) {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/ucum-%s",
                "name": "%sWithUcumUnit",
                "$uses": ["JSONStructureUnits"],
                "type": "%s",
                "ucumUnit": "m"
            }
            """.formatted(typeName, typeName, typeName);

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
        assertThat(result.getErrors()).isEmpty();
    }

    @Test
    @DisplayName("Invalid non-numeric type with ucumUnit")
    void invalidNonNumericTypeWithUcumUnit() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/ucum-string",
                "name": "TextWithUnit",
                "$uses": ["JSONStructureUnits"],
                "type": "string",
                "ucumUnit": "m"
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isFalse();
    }

    @Test
    @DisplayName("Invalid non-string ucumUnit values")
    void invalidNonStringUcumUnitValues() {
        String[] schemas = {
            """
                {
                    "$schema": "https://json-structure.org/meta/extended/v0/#",
                    "$id": "https://test.example.com/schema/ucum-number-value",
                    "name": "NumericUcumUnit",
                    "$uses": ["JSONStructureUnits"],
                    "type": "number",
                    "ucumUnit": 42
                }
                """,
            """
                {
                    "$schema": "https://json-structure.org/meta/extended/v0/#",
                    "$id": "https://test.example.com/schema/ucum-array-value",
                    "name": "ArrayUcumUnit",
                    "$uses": ["JSONStructureUnits"],
                    "type": "number",
                    "ucumUnit": ["m"]
                }
                """,
            """
                {
                    "$schema": "https://json-structure.org/meta/extended/v0/#",
                    "$id": "https://test.example.com/schema/ucum-object-value",
                    "name": "ObjectUcumUnit",
                    "$uses": ["JSONStructureUnits"],
                    "type": "number",
                    "ucumUnit": {"code": "m"}
                }
                """
        };

        for (String schema : schemas) {
            ValidationResult result = validator.validate(schema);
            assertThat(result.isValid()).isFalse();
        }
    }

    @Test
    @DisplayName("Valid object identity array")
    void validObjectIdentityArray() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/relations-identity",
                "name": "OrderIdentity",
                "$uses": ["JSONStructureRelations"],
                "type": "object",
                "properties": {
                    "id": { "type": "string" },
                    "tenantId": { "type": "string" }
                },
                "identity": ["id", "tenantId"]
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
        assertThat(result.getErrors()).isEmpty();
    }

    @Test
    @DisplayName("Valid relations declarations")
    void validRelationsDeclarations() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/relations-declarations",
                "name": "OrderRelations",
                "$uses": ["JSONStructureRelations"],
                "type": "object",
                "properties": {
                    "id": { "type": "string" },
                    "customerId": { "type": "string" }
                },
                "relations": {
                    "customer": {
                        "cardinality": "single",
                        "targettype": { "$ref": "#/definitions/Customer" }
                    }
                },
                "definitions": {
                    "Customer": {
                        "name": "Customer",
                        "type": "object",
                        "properties": {
                            "id": { "type": "string" }
                        }
                    }
                }
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
        assertThat(result.getErrors()).isEmpty();
    }

    @Test
    @DisplayName("Valid single cardinality relation with targettype ref")
    void validSingleCardinalityRelationWithTargettypeRef() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/relations-single",
                "name": "OrderRelations",
                "$uses": ["JSONStructureRelations"],
                "type": "object",
                "properties": {
                    "id": { "type": "string" },
                    "customerId": { "type": "string" }
                },
                "relations": {
                    "customer": {
                        "cardinality": "single",
                        "targettype": { "$ref": "#/definitions/Customer" }
                    }
                },
                "definitions": {
                    "Customer": {
                        "name": "Customer",
                        "type": "object",
                        "properties": {
                            "id": { "type": "string" }
                        }
                    }
                }
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
    }

    @Test
    @DisplayName("Valid multiple cardinality relation with scope")
    void validMultipleCardinalityRelationWithScope() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/relations-multiple",
                "name": "OrderRelations",
                "$uses": ["JSONStructureRelations"],
                "type": "object",
                "properties": {
                    "id": { "type": "string" },
                    "itemIds": { "type": "array", "items": { "type": "string" } }
                },
                "relations": {
                    "items": {
                        "cardinality": "multiple",
                        "targettype": { "$ref": "#/definitions/Item" },
                        "scope": "line-items"
                    }
                },
                "definitions": {
                    "Item": {
                        "name": "Item",
                        "type": "object",
                        "properties": {
                            "id": { "type": "string" }
                        }
                    }
                }
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
    }

    @Test
    @DisplayName("Valid relation with qualifiertype")
    void validRelationWithQualifierType() {
        String schema = """
            {
                "$schema": "https://json-structure.org/meta/extended/v0/#",
                "$id": "https://test.example.com/schema/relations-qualifier",
                "name": "OrderRelations",
                "$uses": ["JSONStructureRelations"],
                "type": "object",
                "properties": {
                    "id": { "type": "string" },
                    "qualifier": { "type": "string" }
                },
                "relations": {
                    "qualifiedCustomer": {
                        "cardinality": "single",
                        "targettype": { "$ref": "#/definitions/Customer" },
                        "qualifiertype": { "$ref": "#/definitions/RelationQualifier" }
                    }
                },
                "definitions": {
                    "Customer": {
                        "name": "Customer",
                        "type": "object",
                        "properties": {
                            "id": { "type": "string" }
                        }
                    },
                    "RelationQualifier": {
                        "name": "RelationQualifier",
                        "type": "string"
                    }
                }
            }
            """;

        ValidationResult result = validator.validate(schema);
        assertThat(result.isValid()).isTrue();
    }

    @Test
    @DisplayName("Invalid Relations extension schemas")
    void invalidRelationsSchemas() {
        String[] schemas = {
            """
                {
                    "$schema": "https://json-structure.org/meta/extended/v0/#",
                    "$id": "https://test.example.com/schema/relations-identity-non-object",
                    "name": "IdentityOnString",
                    "$uses": ["JSONStructureRelations"],
                    "type": "string",
                    "identity": ["id"]
                }
                """,
            """
                {
                    "$schema": "https://json-structure.org/meta/extended/v0/#",
                    "$id": "https://test.example.com/schema/relations-invalid-cardinality",
                    "name": "InvalidCardinality",
                    "$uses": ["JSONStructureRelations"],
                    "type": "object",
                    "properties": {
                        "id": { "type": "string" }
                    },
                    "relations": {
                        "customer": {
                            "cardinality": "many",
                            "targettype": { "$ref": "#/definitions/Customer" }
                        }
                    },
                    "definitions": {
                        "Customer": {
                            "name": "Customer",
                            "type": "object",
                            "properties": {
                                "id": { "type": "string" }
                            }
                        }
                    }
                }
                """
        };

        for (String schema : schemas) {
            ValidationResult result = validator.validate(schema);
            assertThat(result.isValid()).isFalse();
        }
    }
}
